### ISSUE TITLE
UNIMPLEMENTED; only dense, row-major layouts currently supported

### ISSUE THREAD

**Issue by silvasean** [openxla/openxla-pjrt-plugin#20]
## Expected Behavior

I expect the following program to work

```
import jax.numpy as jnp
import numpy as np
a = np.ones((3, 5), dtype=np.float32).T
print(a.shape, a.strides)
jnp.array(a)
```

## Actual Behavior

It crashes with

```
jaxlib.xla_extension.XlaRuntimeError: UNIMPLEMENTED: iree/integrations/pjrt/common/api_impl.cc:697: UNIMPLEMENTED; only dense, row-major layouts currently supported
```

I see we have a TODO for this:

https://github.com/openxla/openxla-pjrt-plugin/blob/2ee66e0c0a8c4044281ca43015703eefff4ea958/iree/integrations/pjrt/common/api_impl.cc#L693

Is that still the recommended solution? It seems like this could possibly be handled much more simply at the Python level.

Also note that functions like [numpy.lib.stride_tricks.sliding_window_view](https://numpy.org/doc/stable/reference/generated/numpy.lib.stride_tricks.sliding_window_view.html) can create more exotic layouts than simple transposes, but currently in IREE we have no way to represent this without a memory blowup.

@stellaraccident for next steps on this

## Steps to Reproduce the Problem

1. Run the snippet above with the plugin.

## Specifications

- Version: HEAD
- Platform: Ubtuntu 22.04

**Comment by stellaraccident** [comment]
When I looked through the existing implementations of the plugin interface, I saw that most were just doing some kind of local transpose to handle this case (with a TODO to do something better). Agreed that this could/should have been better advertised and handled at the Python level. We can have that discussion (i.e. with Skye on the PJRT team) or we can do what everyone else did and add some kind of local swizzle.

If this was going to be load bearing for performance (which is unclear), in my mind, we would implement a utility to generate transpose-modules on the fly and schedule an additional call to perform the layout change. That is still theoretically just a partial implementation, since there is no general guarantee that a strided layout only needs transposes to normalize, but I didn't see evidence that any other backends were being more adventurous than this.

I think that we should implement some kind of basic linearization fallback here, and then also raise this with the PJRT team.

**Comment by silvasean** [comment]
Note to self: to work around this, insert a `.copy()` in the appropriate place. For AlphaFold it is a `.copy()` [here](https://github.com/deepmind/alphafold/blob/2e6a78f08b3a31945ad0fad77f7cf4344ae73d13/alphafold/model/utils.py#L102)

**Comment by skye** [comment]
cc @hawkinsp as well.

In order to do this in Python or in general the PJRT caller, the caller would need to choose the device layout to swizzle to, which may vary across platforms (e.g. TPU uses tiled layout that sometimes requires extra padding). I actually would like to add caller-specified device layouts to jax to enable a broader range of device layouts, but only as an opt-in option, with the default behavior to still let the implementation choose the device layout.

Unless you're saying the caller should first transpose to row-major and then the impl can do whatever it wants? I think I would prefer adding such a helper method to https://github.com/openxla/xla/blob/main/xla/pjrt/c/pjrt_c_api_helpers.h or somewhere so implementations can easily do this, while still leaving open the possibility of doing something more efficient.

So basically I think y'all should implement the swizzle in the plugin :) And possibly add a helper method to avoid code duplication (maybe it could handle device layouts too!).

**Comment by hawkinsp** [comment]
Indeed, NumPy can and will hand you data with arbitrary input byte strides, and the PJRT API must accept them. It's an important memory bandwidth optimization, at least on TPU where we have to transform the data into a tiled layout, that we only read and write the data once.

In the current PJRT implementations we use this code which implements a reasonably high performance CPU transpose.
https://github.com/openxla/xla/blob/main/xla/pjrt/transpose.h

It is largely self-contained and you could reuse it.

You could also perhaps just emit code for transposes (@nicolasvasilache has experimented with that, I know).

**Comment by stellaraccident** [comment]
> In the current PJRT implementations we use this code which implements a reasonably high performance CPU transpose.
> https://github.com/openxla/xla/blob/main/xla/pjrt/transpose.h

Thanks - I had found that when doing the initial implementation and intended to use it. We should probably eventually emit code to transpose, but can do that later.

**Comment by silvasean** [comment]
Reopening this because I don't see anything in https://github.com/openxla/openxla-pjrt-plugin/pull/38 that handles the transposed case (or tests it).

**Comment by ramiro050** [comment]
I will be working on this issue.

**Comment by ramiro050** [comment]
Draft patch: https://github.com/openxla/openxla-pjrt-plugin/pull/199

**Comment by ramiro050** [comment]
Patch https://github.com/openxla/openxla-pjrt-plugin/pull/199 has landed.
